waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java

```diff
@@ -252,62 +255,90 @@ private boolean isWhitelisted(String databasePrefix, String database) {
+
         for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsByDbPattern(databasePattern)
             .entrySet()) {
-          try {
-            DatabaseMapping mapping = mappingWithPattern.getKey();
-            String pattern = mappingWithPattern.getValue();
-            List<String> databases = mapping.getClient().get_databases(pattern);
-            for (String database : databases) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
+          Runnable runnableTask = () -> {
+            try {
+              DatabaseMapping mapping = mappingWithPattern.getKey();
+              String pattern = mappingWithPattern.getValue();
+              List<String> databases = mapping.getClient().get_databases(pattern);
+              for (String database : databases) {
+                if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
+                  combined.add(mapping.transformOutboundDatabaseName(database));
+                }
               }
+            } catch (TException e) {
+              LOG.warn("Can't fetch databases by pattern: {}", e.getMessage());
             }
-          } catch (TException e) {
-            LOG.warn("Can't fetch databases by pattern: {}", e.getMessage());
-          }
+          };
+          executorService.submit(runnableTask);
         }
+        shutdownExecutorService(executorService);
         return combined;
       }
 
       @Override
       public List<String> getAllDatabases() {
         List<String> combined = new ArrayList<>();
-        for (DatabaseMapping mapping : databaseMappings()) {
-          try {
-            List<String> databases = mapping.getClient().get_all_databases();
-            for (String database : databases) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        for (final DatabaseMapping mapping : databaseMappings()) {
+          Runnable runnableTask = () -> {
+            try {
+              List<String> databases = mapping.getClient().get_all_databases();
+              for (String database : databases) {
+                if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
+                  combined.add(mapping.transformOutboundDatabaseName(database));
+                }
               }
+              LOG.info("Finished executing getAllDatabases for " + mapping.getDatabasePrefix());
+            } catch (TException e) {
+              LOG.warn("Can't fetch databases: {}", e.getMessage());
             }
-          } catch (TException e) {
-            LOG.warn("Can't fetch databases: {}", e.getMessage());
-          }
+          };
+          executorService.submit(runnableTask);
```

Comment:
I think when you run this you'll see that the order of the database names coming back is not fixed anymore. This will be very confusing for a user. I think we need to have the runnable task return a results. Then loop over the Futures that are returned in order and collect the results. This way we process in parallel but returns results always in the same order.

Reply:
Ah, I see. This also means that I don't need to change a lot of the tests anymore.

After I loop over the Futures, I still need to shutdown the ExecutorService, right?

Reply:
yes that's good practise. To wait for a future you need to call get() I think there is a method that takes a timeout which we should use.